File: src/cpp/mkhead.cpp
/**************************************************************
 * File Name   : mkhead.cpp
 * Author      : ThreeDog
 * Date        : Thu Nov  3 11:05:14 2016
 * Description : 快捷命令 制作.h文件
 *
 **************************************************************/

#include <iostream>
#include <cstring>
#include <fstream>
#include <cstdlib>
using namespace std;

int main(int argc,char **argv){
	int size = 20;
	if(argc < 2){
		cerr << "arguement error: 命令mkhead需要包含一个参数 " << endl;
		return 1;
	}
	else if(argc > 2){
		cerr << "arguement error: 命令mkhead只能包含一个参数 " << endl;
		return 1;
	}


	ofstream output_file(argv[1],ios::trunc);
	if(!output_file){
		cerr << "open file " << argv[1] << "fail" << endl;
		return 3;
		}
	char *p = new char[size];
	char *a = new char[size];
	char *name = new char[size];
	strcpy(name,argv[1]);
	strcpy(a,"vi ");
	char *n = strtok(argv[1],".");

	while(strlen(n) > size){
		char *q = new char[size+10];
		delete[] p;
		p = q;
		char *r = new char[size+10];
		strcpy(r,a);
		delete[] a;
		a = r;
		size += 10;
		}
	strcpy(p,n);
	strcat(a,name);
	for(int i = 0; i < size; i++)
		*(p+i) = toupper(*(p+i));
	char buff[size+50];
	memset(buff,0,sizeof(buff));
	strcat(buff,"#ifndef _");
	strcat(buff,p);
	strcat(buff,"_H_ \n");
	strcat(buff,"#define _");
	strcat(buff,p);
	strcat(buff,"_H_ \n\n\n\n\n\n\n");
	strcat(buff,"#endif");

	time_t now_time;
    now_time = time(NULL);
	output_file << "/**************************************************************"<<endl;
	output_file << " * File Name   : " << argv[1] << ".h" << endl;
	output_file << " * Author      : ThreeDog"<<endl;
	output_file << " * Date        : " << ctime(&now_time);
	output_file << " * Description : " << endl << " *" << endl;
	output_file << " **************************************************************/"<<endl;

	output_file.write(buff,strlen(buff));
	output_file.close();
	system(a);
	delete[] name;
	return 0;
}

File: src/golang/mkhead.go
/**************************************************************
* File Name   : mkhead.go
* Author      : ThreeDog
* Date        : 2019-12-23 20:04:41
* Description : 
*
**************************************************************/

package main

import (
	"fmt"
	"os"
	"log"
	"bytes"
	"time"
	"strings"
)

func PathExists(path string) (bool, error) {
	_, err := os.Stat(path)
	if err == nil {
		return true, nil
	}
	if os.IsNotExist(err) {
		return false, nil
	}
	return false, err
}

/**
 检查错误
 */
func checkErr(err error){
	if err!=nil{
		log.Fatal(err)
	}
}

func main()  {
	// 命令行参数获取
	if len(os.Args) != 2 {
		fmt.Printf("arguement error: 命令mkc需要有且只有一个参数 \n")
		return 
	}
	
	res,_ := PathExists(os.Args[1])
	// 如果文件已经存在，替换一下文件名 
	if res {
		err := os.Rename(os.Args[1],os.Args[1]+".bak")
		if err != nil{
			fmt.Printf("文件%q已存在，将文件重命名为%q.bak时失败 \n",os.Args[1],os.Args[1])
			return 
		}
	}
	
	// 打开文件
	file,err:=os.OpenFile(os.Args[1],os.O_CREATE|os.O_APPEND|os.O_WRONLY,0644)
	checkErr(err)
	var buffer bytes.Buffer
	buffer.WriteString("/**************************************************************\n")
	buffer.WriteString("* File Name   : " + os.Args[1] + "\n")
	buffer.WriteString("* Author      : ThreeDog\n")
	buffer.WriteString("* Date        : "+time.Now().Format("2006-01-02 15:04:05")+"\n")
	buffer.WriteString("* Description : \n")
	buffer.WriteString("*\n")
	buffer.WriteString("**************************************************************/\n\n")
	
	var name string;
	name = strings.ToUpper(os.Args[1])
	name = strings.Replace(name, ".", "_", -1 )

	buffer.WriteString("#ifndef __"+name+"__\n")
	buffer.WriteString("#define __"+name+"__\n\n\n\n\n\n\n\n\n\n")
	buffer.WriteString("#endif  // __"+name+"__\n")

	file.Write([]byte(buffer.String()))
	file.Close()
}
